crates/holochain_state/src/source_chain.rs

```diff
@@ -626,113 +626,117 @@ where
-                                            Some(CapGrant::RemoteAgent(
-                                                acc_zome_call_cap_grant,
-                                            )) => {
-                                                match acc_zome_call_cap_grant.access {
-                                                    // an assigned acc takes precedence
-                                                    CapAccess::Assigned { .. } => acc,
-                                                    // transferable acc takes precedence
-                                                    CapAccess::Transferable { .. } => acc,
-                                                    // current grant takes preference over other accs
-                                                    _ => Some(grant),
-                                                }
-                                            }
-                                            None => Some(grant),
-                                            // authorship should be short circuited and filtered by now
-                                            _ => unreachable!(),
-                                        },
-                                        CapAccess::Unrestricted => match acc {
-                                            Some(_) => acc,
-                                            None => Some(grant),
-                                        },
-                                    }
-                                }
-                                // ChainAuthor should have short circuited and be filtered out already
-                                _ => unreachable!(),
-                            };
-                            Ok(acc)
-                        },
-                    )
+            .async_reader({
+                let agent_pubkey = self.agent_pubkey().clone();
+                move |txn| -> Result<_, DatabaseError> {
+                    // prepare sql statement depending on provided cap secret
+                    let mut statement = match &check_secret {
+                        Some(cap_secret) => {
+                            // cap grant for cap secret must exist
+                            // that has not been updated or deleted
+                            let sql = "
+                                SELECT Entry.blob
+                                FROM Entry
+                                WHERE Entry.cap_secret = ?1
+                                AND (
+                                    SELECT COUNT(Action.hash)
+                                    FROM Action
+                                    WHERE Action.author = ?2
+                                    AND (
+                                        Action.original_entry_hash = Entry.hash
+                                        OR Action.deletes_entry_hash = Entry.hash
+                                    )
+                                ) = 0
```

Comment:
If we are able to move away from string-manipulated sql, I'd love it if we could put the queries in separate `.sql` files in the holochain_sqlite crate so they can be linted.